Added support for multiple feeds (i.e. generating both Atom and RSS)

LunarEclipse363 opened the pull request:
## Introduction
As requested many times before, for example:
- https://zola.discourse.group/t/generate-both-rss-and-atom-feeds/441
- #2083

## Potential improvements
~~I don't know if/how Zola handles deprecating config options, if there's anything I should add to let the user know the feed related options changed please let me know how to do it.~~
Implemented backwards-compatibility as suggested by @selfisekai 

## Formalities
**Sanity check:**

* [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/getzola/zola/pulls) for the same update/change?

**Code changes**
* [x] Are you doing the PR on the `next` branch?

**If the change is a new feature or adding to/changing an existing one:**
* [x] Have you created/updated the relevant documentation page(s)?

LunarEclipse363 committed:
Added support for multiple feeds

selfisekai commented on components/config/src/config/languages.rs:
```diff
@@ -14,11 +14,11 @@ pub struct LanguageOptions {
-    /// Defaults to "atom.xml", with "rss.xml" also having a template provided out of the box.
-    pub feed_filename: String,
+    /// Whether to generate feeds for that language, defaults to `false`
+    pub generate_feeds: bool,
```
this will be a breaking change, imo should be marked `#[serde(alias = "generate_feed")]` for backwards compat

selfisekai commented on components/config/src/config/languages.rs:
```diff
@@ -14,11 +14,11 @@ pub struct LanguageOptions {
+    pub generate_feeds: bool,
+    /// The filenames to use for feeds. Used to find the templates, too.
+    /// Defaults to ["atom.xml"], with "rss.xml" also having a template provided out of the box.
+    pub feed_filenames: Vec<String>,
```
`feed_filename = "rss.xml"` should still be accepted too for compat. in another project i've solved this with a `serde(deserialize_with)` macro: https://codeberg.org/selfisekai/electron_tasje/src/commit/3d34593c358a1d68741f3c3d0554447f33d9ec16/src/config.rs#L74-L102

LunarEclipse363 committed:
Implemented backwards-compatibility for feed config

LunarEclipse363 committed:
Added a test for feed config backwards-compat, fixed bugs

- Fixed language config merge bug found by a test
- Adjusted two existing tests to fully check stuff related to multiple feeds
- Added a new test for backwards-compatibility of the changes
- Fixed bugs found by the newly added test

Keats commented on components/config/src/config/languages.rs:
```diff
@@ -14,11 +15,13 @@ pub struct LanguageOptions {
+    pub generate_feeds: bool,
+    /// The filenames to use for feeds. Used to find the templates, too.
+    /// Defaults to ["atom.xml"], with "rss.xml" also having a template provided out of the box.
+    #[serde(alias = "feed_filename", deserialize_with = "might_be_single")]
```
the `might_be_single` name is not great

Keats commented on components/config/src/config/mod.rs:
```diff
@@ -395,6 +400,48 @@ impl Default for Config {
+
+#[derive(Debug, Clone, Deserialize, PartialEq, Eq)]
+#[serde(untagged)]
+pub(crate) enum MightBeSingle<T> {
```
If we do support that, we need to add a deprecation message when there is only one value. I'm not entirely we need to though, it's a pretty simple change to do from the user POV.

As for name... `SingleOrVec`?

LunarEclipse363 replied:
The problem I ran into while testing before adding the backwards-compatibility is that Zola doesn't error out if you leave the old value names in the config file, instead using default values, which is a very bad way to handle it. The user will think they have the configuration option set while it's not until they realize it's broken purely due to a single-letter rename.

I can change the name and add a deprecation message, sure.

LunarEclipse363 replied:
Update: adding a meaningful deprecation message seems to be much harder than I expected due to the constraints of how serde works

Keats commented on components/content/src/front_matter/section.rs:
```diff
@@ -68,8 +68,8 @@ pub struct SectionFrontMatter {
     /// Whether to generate a feed for the current section
-    #[serde(skip_serializing)]
-    pub generate_feed: bool,
+    #[serde(skip_serializing, alias = "generate_feed")]
```
you don't need it here i think

LunarEclipse363 replied:
Zola will ignore unknown fields here as well though which again ends up with the "suddenly an option stopped working without an error" problem

Which I'm not sure is solvable within serde

Keats commented on components/site/src/feeds.rs:
```diff
@@ -73,18 +73,23 @@ pub fn render_feed(
-    };
+    let mut feeds = Vec::new();
+    for feed_filename in &site.config.feed_filenames {
+        let mut context = context.clone();
```
That's pretty wasteful.

Can you render them one by one and just updating `context.insert("feed_url", &feed_url);` in the loop? I can't remember what `additional_context_fn(context);` does though

LunarEclipse363 replied:
`render_template()` takes ownership of the context so I need to clone it on every loop regardless

LunarEclipse363 committed:
Renamed MightBeSingle to SingleOrVec

LunarEclipse363 committed:
Made the multiple feeds config changes "loudly" backwards-incompatible

Keats commented on components/content/src/front_matter/section.rs:
```diff
@@ -67,13 +68,23 @@ pub struct SectionFrontMatter {
     pub aliases: Vec<String>,
+    /// Deprecated
+    #[serde(skip_serializing)]
+    pub generate_feed: Deprecated<DeprecatedGenerateFeed>,
```
:+1:

LunarEclipse363 committed:
added #[serde(deny_unknown_fields)] to front-matter, fixed problems this found in tests

Keats reviewed (approved):
Nice, thanks
